File: src/main/scala/org/helianto/security/service/AbstractDetailsService.java
package org.helianto.security.service;

import org.helianto.security.domain.Secret;
import org.helianto.security.repository.SecretRepository;
import org.helianto.user.repository.UserProjection;
import org.helianto.user.repository.UserRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.dao.DataAccessException;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

/**
 * Base class to security details services.
 * 
 * @author mauriciofernandesdecastro
 */
public class AbstractDetailsService {

    protected static Logger logger = LoggerFactory.getLogger(AbstractDetailsService.class);
    
	@Autowired
    private SecretRepository secretRepository;

	@Autowired
	protected UserRepository userGroupRepository;

	/**
	 * Step 1: retrieve the identity secret.
	 * 
	 * @param identityKey
	 *
	 * @throws UsernameNotFoundException
	 * @throws DataAccessException
	 */
	@Transactional
	protected Secret loadIdentitySecretByKey(String identityKey) throws UsernameNotFoundException, DataAccessException {

		Secret secret = secretRepository.findByIdentityKeyOrEmail(identityKey);

		if (secret !=null) {
			return secret;
		}
		
		logger.error("Unable to load by user name with {}.", identityKey);
		throw new UsernameNotFoundException("Unable to find user name for "+identityKey);
		
	}
	
	/**
	 * Step 2: list users given the identity.
	 * 
	 * @param identityId
	 */
	protected List<UserProjection> loadUserListByIdentityId(int identityId) {
		List<UserProjection> userList = userGroupRepository.findByIdentity_IdOrderByLastEventDesc(identityId);
		
		if (userList!=null && userList.size()>0) {
			logger.debug("Found {} user(s) matching id={}.", userList.size(), identityId);
			return userList;
		}

		logger.error("Unable to load by user list");
		throw new UsernameNotFoundException("Unable to find any user for identity id "+identityId);
	}
	
}

File: src/main/scala/org/helianto/security/service/AbstractDetailsService.scala
package org.helianto.security.service

import org.helianto.core.service.ContextInstallService
import org.helianto.security.domain.{UserAuthority, UserDetailsAdapter}
import org.helianto.security.repository.{SecretRepository, UserAuthorityRepository}
import org.helianto.user.repository.{UserProjection, UserRepository}
import org.slf4j.{Logger, LoggerFactory}
import org.springframework.beans.factory.annotation.{Autowired, Value}
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UsernameNotFoundException

class AbstractDetailsService {

  @Value("${helianto.contextName}") val contextName: String = ""
  @Autowired val secretRepository: SecretRepository = null
  @Autowired val userRepository: UserRepository = null
  @Autowired val userAuthorityRepository: UserAuthorityRepository = null
  @Autowired val contextInstallService: ContextInstallService = null

  val logger: Logger = LoggerFactory.getLogger(classOf[AbstractDetailsService])

  private[service]  def loadIdentitySecretByKey(identityKey: String) =
    Option(secretRepository.findByIdentityKeyOrEmail(identityKey)) match {
      case Some(s) => s
      case None => error(identityKey)
    }

  private[service]  def loadLastUserByIdentityId(identityId: String): UserProjection =
    Option(userRepository.findByIdentity_IdOrderByLastEventDesc(identityId)) match {
      case Some(list) if !list.isEmpty => list.get(0)
      case _ => error(identityId)
    }

  private[service] def updateLastEvent(userDetails: UserDetailsAdapter): Unit = {
    Option(userRepository.findOne(userDetails.getUserId)) match {
      case Some(u) =>
        userRepository.saveAndFlush(u.updateLastEvent())
      case _ => error(userDetails.getUserId)
    }
  }

  private[service] def updateAuthorities(user: UserDetailsAdapter) =
    Option(userAuthorityRepository.findByUserIdOrderByServiceCodeAsc(user.getUserId)) match {
      case Some(userAuthorities) =>
        import collection.JavaConversions._
        val authorities = userAuthorities
          .flatMap(getUserAuthoritiesAsString(_))
          .map(new SimpleGrantedAuthority(_))
          .distinct
          .:+(new SimpleGrantedAuthority(s"CONTEXT_${contextName}"))
          .:+(new SimpleGrantedAuthority(s"SELF_ID_${user.getIdentityId}"))
          .:+(new SimpleGrantedAuthority(s"USER_ID_${user.getUserId}"))
          .:+(new SimpleGrantedAuthority(s"ENTITY_ID_${user.getUser.getEntityId}"))
          .toList
        user.updateAuthorities(authorities)
      case None => user
    }

  private[service] def afterLoadUser(userDetails: UserDetailsAdapter) = {
    updateLastEvent(userDetails)
    updateAuthorities(userDetails)
  }

  /**
    * As an additional control, if an error occurs during login, the context installation service is invoked.
    *
    * @param source
    *
    * @throws UsernameNotFoundException
    */
  private[service] def error(source: String) = {
    contextInstallService.installDefaultContext()
    logger.error("Unable to load by user name with {}.", source)
    throw new UsernameNotFoundException("Unable to find user name for " + source)
  }

  private[service] def getUserAuthoritiesAsString(authority: UserAuthority) = {
    expandUserAuthorities(authority.getServiceCode.toUpperCase, authority.getServiceExtension)
  }

  private[service] def expandUserAuthorities(serviceCode: String, extensions: String): Iterable[String] = {
    Option(extensions) match {
      case Some(s) => s.replaceAll(" ", "").split(",").map(_.toUpperCase()).map(e => s"ROLE_${serviceCode}_${e}")
      case None => Seq(s"ROLE_${serviceCode}_")
    }
  }

}